exercises/practice/allergies/.approaches/introduction.md

```diff
@@ -0,0 +1,238 @@
+
+The bit masks (1, 2, 4 ...) are duplicated once, which most people would say is Ok (people commonly refactor code when it is duplicated 3 times or more).
+
+The compiler does not guarantee that all `Allergy` types are represented in the case expression.
+
+[Read more about this approach][bitwise-and-case].
+
+## Approach: Bitwise, List and type alias
+
+This approach uses a type alias to represent the bit position and the allergy.
+
+```elm
+type alias AllergyAndScore = {
+  allergy: Allergy;
+  bitPosition: Int;
+}
+
+allergiesAndScores : List AllergyAndScore
+allergiesAndScores =
+    [ AllergyAndScore Eggs 1
+    , AllergyAndScore Peanuts 2
+    , AllergyAndScore Shellfish 3
+    , AllergyAndScore Strawberries 4
+    , AllergyAndScore Tomatoes 5
+    , AllergyAndScore Chocolate 6
+    , AllergyAndScore Pollen 7
+    , AllergyAndScore Cats 8
+    ]
+
+toList : Int -> List Allergy
+toList score =
+  allergiesAndScores
+    |> List.Map (\allergyAndScore -> ( 2 ^ allergyAndScore.bitPosition, allergyAndScore.allergy ))
+    |> List.filter (\( s, _ ) -> Bitwise.and s score > 0)
+    |> List.map Tuple.second
+```
+
+This approach at is the best at embedding the domain concept of using the bit positions in an `Int` to represent a list of `Allergy`.
+You can also use a tuple instead of a type alias, but then you can't name the variables, so you lose some of this benefit.
+
+The compiler does not guarantee that the `allergies` list contains all the `Allergy` types.
+You can use the [type iterator pattern][type-iterator-pattern] to fix this.
+
+[Read more about this approach][bitwise-and-type-alias].
+
+## Other approaches
+
+There are also more esoteric approaches that don't use `Bitwise`, [such as this one using tuples and recursion][esoteric-approach]. `Bitwise` is the idiomatic way to do bit manipulation in Elm, but other solutions can still be interesting to look at.
+
+## Type iterator pattern
+
+In order to allow the compiler to guarantee that all the `Allergy` types are present in the list (or dict or similar), you can use the type iterator pattern, which involves recursion and a case expression.
+This is explained in  detail on the [elm patterns website][https://sporto.github.io/elm-patterns/basic/type-iterator.html], and you can see how the code would look for this example below.
+
+```elm
+allAllergies : List Allergy
+allAllergies =
+  let
+    next list =
+      case head list of
+        Nothing -> Eggs :: list |> next
+        Just Eggs -> Peanuts :: list |> next
+        Just Peanuts -> Shellfish :: list |> next
+        Just Shellfish -> Strawberries :: list |> next
+        Just Strawberries -> Tomatoes :: list |> next
+        Just Tomatoes -> Chocolate :: list |> next
+        Just Chocolate -> Pollen :: list |> next
+        Just Pollen -> Cats :: list |> next
+        Just Cats -> list
```

Comment:
I'm not sure what you mean by this?

Reply:
`allAllergies` will return the list `[Cats, Pollen, Chocolate, ...]`, but every other time we produce a list of allergies, we define in the the opposite order `[Eggs, Peanuts, Shellfish, ...]`.

It's maybe not a huge deal since this is only meant to illustrate the type iterator pattern, but if someone copy/pastes it in their code, it won't work.